base/linalg/matmul.jl

```diff
@@ -424,7 +426,7 @@ end
 function generic_matmatmul{T,S}(tA, tB, A::AbstractVecOrMat{T}, B::AbstractMatrix{S})
     mA, nA = lapack_size(tA, A)
     mB, nB = lapack_size(tB, B)
-    C = similar(B, promote_op(*, arithtype(T), arithtype(S)), mA, nB)
+    C = similar(B, promote_op(matprod, arithtype(T), arithtype(S)), mA, nB)
```

Comment:
Probably by mistake. But IIUC, `arithtype` is only needed to convert `Bool` to `Int` when doing matrix multiplication, which the `promote_op(matprod, ...)` does even without the `arithtype`, so no worries here, but maybe a chance for cleanup.